In [44]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')

In [4]:
df.head()

,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,model_year,origin,fuel_type,drivetrain,num_doors,fuel_efficiency_mpg
0,170,3.0,159.0,3413.433759,17.7,2003,Europe,Gasoline,All-wheel drive,0.0,13.231729
1,130,5.0,97.0,3149.664934,17.8,2007,USA,Gasoline,Front-wheel drive,0.0,13.688217
2,170,NaN,78.0,3079.038997,15.1,2018,Europe,Gasoline,Front-wheel drive,0.0,14.246341
3,220,4.0,NaN,2542.392402,20.2,2009,USA,Diesel,All-wheel drive,2.0,16.912736
4,210,1.0,140.0,3460.870990,14.4,2009,Europe,Gasoline,All-wheel drive,2.0,12.488369


In [5]:
df.describe()

,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,model_year,num_doors,fuel_efficiency_mpg
count,9704.000000,9222.000000,8996.000000,9704.000000,8774.000000,9704.000000,9202.000000,9704.000000
mean,199.708368,3.962481,149.657292,3001.280993,15.021928,2011.484027,-0.006412,14.985243
std,49.455319,1.999323,29.879555,497.894860,2.510339,6.659808,1.048162,2.556468
min,10.000000,0.000000,37.000000,952.681761,6.000000,2000.000000,-4.000000,6.200971
25%,170.000000,3.000000,130.000000,2666.248985,13.300000,2006.000000,-1.000000,13.267459
50%,200.000000,4.000000,149.000000,2993.226296,15.000000,2012.000000,0.000000,15.006037
75%,230.000000,5.000000,170.000000,3334.957039,16.700000,2017.000000,1.000000,16.707965
max,380.000000,13.000000,271.000000,4739.077089,24.300000,2023.000000,4.000000,25.967222


## Q1. Pandas Version

In [6]:
pd.__version__

'3.0.0'

## Q2. Records Count

In [7]:
df.shape[0]

9704

In [8]:
len(df)

9704

In [9]:
len(df.index)

9704

### There is 9704 records in this dataset

## Q3. Fuel types

In [10]:
df['fuel_type'].unique()

<StringArray>
['Gasoline', 'Diesel']
Length: 2, dtype: str

In [11]:
df['fuel_type'].nunique()

2

## Q4. Missing values

In [12]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [13]:
df.isnull().any()

engine_displacement    False
num_cylinders           True
horsepower              True
vehicle_weight         False
acceleration            True
model_year             False
origin                 False
fuel_type              False
drivetrain             False
num_doors               True
fuel_efficiency_mpg    False
dtype: bool

In [14]:
df.isnull().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

## Q5. Max Fuel efficiency

In [15]:
df.groupby('origin').fuel_efficiency_mpg.max().round(3)

origin
Asia      23.759
Europe    25.967
USA       24.971
Name: fuel_efficiency_mpg, dtype: float64

In [16]:
df[df["origin"] == "Asia"]["fuel_efficiency_mpg"].max()

np.float64(23.759122836520497)

In [17]:
df.loc[df["origin"] == "Asia", "fuel_efficiency_mpg"].max()

np.float64(23.759122836520497)

In [18]:
df.query("origin == 'Asia'")["fuel_efficiency_mpg"].max()

np.float64(23.759122836520497)

## Q6. Median value of horsepower

In [19]:
df['horsepower'].median()

np.float64(149.0)

In [20]:
df["horsepower"].value_counts()

horsepower
152.0    142
145.0    141
151.0    134
148.0    130
141.0    130
        ... 
40.0       1
57.0       1
245.0      1
252.0      1
61.0       1
Name: count, Length: 192, dtype: int64

In [21]:
df['horsepower'].mode()

0    152.0
Name: horsepower, dtype: float64

In [22]:
new = df['horsepower'].fillna(152)

In [23]:
new.median()

np.float64(152.0)

## Q7. Sum of weights

In [52]:
df_Asia = df.query("origin == 'Asia'")
df_Asia.head(7)

,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,model_year,origin,fuel_type,drivetrain,num_doors,fuel_efficiency_mpg
8,250,1.0,174.0,2714.219310,10.3,2016,Asia,Diesel,Front-wheel drive,-1.0,16.823554
12,320,5.0,145.0,2783.868974,15.1,2010,Asia,Diesel,All-wheel drive,1.0,16.175820
14,200,6.0,160.0,3582.687368,14.9,2007,Asia,Diesel,All-wheel drive,0.0,11.871091
20,150,3.0,197.0,2231.808142,18.7,2011,Asia,Gasoline,Front-wheel drive,1.0,18.889083
21,160,4.0,133.0,2659.431451,NaN,2016,Asia,Gasoline,Front-wheel drive,-1.0,16.077730
34,150,2.0,149.0,2844.227534,17.3,2014,Asia,Diesel,Front-wheel drive,0.0,15.353788
38,190,5.0,158.0,3761.994038,NaN,2019,Asia,Gasoline,Front-wheel drive,1.0,11.645724


In [57]:
X_df = df_Asia[['vehicle_weight', 'model_year']].head(7)
X_df

,vehicle_weight,model_year
8,2714.219310,2016
12,2783.868974,2010
14,3582.687368,2007
20,2231.808142,2011
21,2659.431451,2016
34,2844.227534,2014
38,3761.994038,2019


In [59]:
X = X_df.to_numpy()
X

array([[2714.21930965, 2016.        ],
       [2783.86897424, 2010.        ],
       [3582.68736772, 2007.        ],
       [2231.8081416 , 2011.        ],
       [2659.43145076, 2016.        ],
       [2844.22753389, 2014.        ],
       [3761.99403819, 2019.        ]])

In [67]:
XTX = np.dot(X.T, X)
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [69]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [70]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [72]:
w = np.dot((np.dot(XTX_inv, X.T)), y)
w

array([0.01386421, 0.5049067 ])

In [73]:
w.sum()

np.float64(0.5187709081074016)